Add transposedConv2D (#596)

t-ae opened the pull request:
`TansposedConv2D` layer is using internal API `conv2DBackpropInput`.
It makes creating custom transposed convolutional layers outside the library hard.

This PR exposes that API with new name `transposedConv2D`.

Shashi456 commented on Sources/TensorFlow/Operators/NN.swift:
```diff
@@ -142,46 +142,56 @@ func _vjpConv2D<Scalar: TensorFlowFloatingPoint>(
-@usableFromInline
-func conv2DBackpropInput<Scalar: TensorFlowFloatingPoint>(
-    _ x: Tensor<Scalar>,
+/// Returns a 2-D transposed convolution with the specified input, filter, strides, and padding.
```
I'm not really for this renaming of gradient helper for the conv operation. It would be better if a separate wrapper for `transposedConv2D` is written which calls `conv2DBackpropInput`. This current renaming would also confuse people who check the source code about as to why transposed functions are being called for a convolution operation.

t-ae replied:
I originally proposed to add alias for it in #596, and @dan-zheng proposed to just rename it.
Adding alias is less confusing for today's contributors, but it won't have much meaning for future contributors, so I decided to take rename way.

Shashi456 replied:
The Tensorflow `Conv2DTranspose` calls the generated op as a wrapper, [here](https://github.com/tensorflow/tensorflow/blob/964a78f3834a17b44fb8898f3d3be95883fa0007/tensorflow/python/ops/nn_ops.py#L2218). Similarly done for `conv1dTranspose` and `conv3dTranpose`, While the conv2d [calls](https://github.com/tensorflow/tensorflow/blob/964a78f3834a17b44fb8898f3d3be95883fa0007/tensorflow/python/ops/nn_ops.py#L925), `input` and `filter` separately. 
A separation of concerns, with separate functions make sense and make for a cleaner api, than renaming `conv2dBackpropInput` as `transposedConv2D`.

t-ae replied:
And there's separate `conv2d_backprop_input` [here](https://github.com/tensorflow/tensorflow/blob/964a78f3834a17b44fb8898f3d3be95883fa0007/tensorflow/python/ops/nn_ops.py#L2078).
`conv2d_transpose_v2` has extra operations (name_scope and data_format part), so they necessarily should be separated.
In S4TF there's no such operations. I'm not sure yet we should separate them.

Shashi456 replied:
The `name_scope` and `data_format` isn't exactly an operation as they are related to input type. We currently only support the NHWC format, I think once we are thoroughly done implementing layers, We should start looking to support the NCWH format.

Shashi456 commented on Sources/TensorFlow/Operators/NN.swift:
```diff
@@ -142,46 +142,56 @@ func _vjpConv2D<Scalar: TensorFlowFloatingPoint>(
         inputSizes: shape,
         filter: filter,
-        outBackprop: x,
+        outBackprop: input,
```
Do tests pass after this change? Afaik all convolution operations are checked with their gradients as well.

t-ae replied:
It's merely a renaming of argument name.
I thought it should have same signature as `conv2D`.
Of course tests are passing.

Shashi456 replied:
@t-ae oh sorry, overlooked the renaming!

t-ae commented:
I restored `conv2DBackpropInput` and added `transposedConv2D`.
Now `TransposedConv*D` layers use `transposedConv2D` and `_vjpConv2D*` funcs use `conv2DBackpropInput`.

> Could you please add some tests for transposedConv2D and its derivative? You can adapt existing tests for conv2D.

It looks there's no tests for `conv2D`. But there are tests for `Conv2D` instead.
Also `TransposedConv2D` has its test, but it didn't have test of its gradients, so I just added it.


By the way I got two questions.

1. `_vjpConv2DBackpropInput` and some other functions are `@usableFromInline`.
Is there any reason to prefer `@usableFromInline` to `@inlinable`?

2. `transposedConv2D` has argument `shape: Tensor<Int32>`. It seems we can change its type to `TensorShape`.
Do you think it's good idea?

Shashi456 commented:
@t-ae all shape variables in the operators file for convolutional operator are `Tensor<Int32>` because the backend generated ops need them in that format, so making it a `TensorShape` is not a good idea.

t-ae commented:
> all shape variables in the operators file for convolutional operator are Tensor<Int32> 

They are internal but `transposedConv2D` is public.
We'd better decide its signature from the point of view of users.

According to [tf.nn.conv2d_transpose](https://www.tensorflow.org/api_docs/python/tf/nn/conv2d_transpose) documentation,

> output_shape: A 1-D Tensor representing the output shape of the deconvolution op.

`TensorShape` is always 1-D so user can't pass wrong parameter.

I don't know if there's case rank of `shape` is not 4. If there's no such case, it should be rather `(Int, Int, Int, Int)` than `TensorShape`.

t-ae commented:
@dan-zheng 
The discussion about the type of `shape` is left over.
Do you have any comments?
